File: c/insertion.c
#include<stdio.h>
#include<stdlib.h>
#include<time.h>

#define REPETITION 22
#define SIZE 10000
void insertion(int A[], int min, int max)
{
    int i, j, key;
    for (j=1; j<=max; j++)
    {
        key = A[j];
        i = j-1;
        while(i>=0 && A[i]>key)
        {
            A[i+1] = A[i];
            i--;
        }
        A[i+1] = key;
    }
}

int * readNumbers(char *file, int size)
{
    int i=0, aux,*vet = malloc(sizeof(int) * size);

    FILE *f = fopen(file, "r");
    while(EOF!=fscanf(f, "%d", &aux) && i<size)
        vet[i++] = aux;
    fclose(f);
    return vet;
}


int * orderNumbers(int size)
{
    int i=0, aux, *vet = malloc(sizeof(int) * size);
    while(i<size)
    {
        vet[i] = i++;
    }
    return vet;
}

int * reverseOrderNumbers(int size)
{
    int i=0, aux=size, *vet = malloc(sizeof(int) * size);
    while(i<size)
    {
        vet[i++] = aux--;
    }
    return vet;
}

void execute(int qtd, char order)
{
    int i=0, *vet;
    double time_spent;
    clock_t end, begin;
    printf("INSERTION SORT (Size=%d, rept=%d) %c\n", qtd, REPETITION, order);
    for(i=0; i<REPETITION; i++)
    {
        if(order == 'R')
            vet = reverseOrderNumbers(qtd);
        else if(order == 'O')
            vet = orderNumbers(qtd);
        else if(order == 'F')
            vet = readNumbers("random_numbers.txt", qtd);
        else
            return;

        begin = clock();
        insertion(vet, 0, qtd-1);
        end = clock();

        time_spent = ((double)(end - begin) / CLOCKS_PER_SEC) * 1000;
        printf("%.f\n", time_spent);
        free(vet);
    }
}


int main()
{
   // execute(200000, 'O');
   // execute(200000, 'F');
   // execute(200000, 'R');

    //execute(500000, 'O');
    //execute(500000, 'F');
    execute(1000000, 'R');
//    execute(500000, 'R');
//    execute(800000, 'R');
//    execute(1000000, 'R');

    return 0;
}

File: python/insertion.py
import time

def readNumbers(file, size):
    vet = []
    i = 0
    f = open(file, "r")
    for line in f:
        if i > size:
            break
        vet.append(int(line))
        i = i+1
    f.close()
    return vet

def insertion(A):  
    # through 1 to len(A) 
    for i in range(1, len(A)): 
        key = A[i] 
        j = i-1
        while j >=0 and key < A[j] : 
                A[j+1] = A[j] 
                j -= 1
        A[j+1] = key 

def orderNumbers(size):
    i = 0
    vet = []
    while i<size:
        vet.append(i)
        i+=1
    return vet

def reverseOrderNumbers(size):
    i = 0
    vet = []
    while size>0 :
        vet.append(size)
        size-=1
    return vet;

# main
REPETITION = 20
SIZE = 200000
vet = []
print("INSERTION SORT (Size=" + str(SIZE) + ", rept="+ str(REPETITION) + ")")
for x in range(REPETITION):
    vet = reverseOrderNumbers(SIZE)
#readNumbers("random_numbers.txt", SIZE)
    
    
    begin = int(round(time.time() * 1000))
    insertion(vet)
    end = int(round(time.time() * 1000))

    print (end-begin)
#print ("Sorted array is:") 
#for i in range(len(vet)):
#    print ("%d" %vet[i])
